### PR TITLE
ZeroCopy: All of `zero_copy::text`

### PR DESCRIPTION
**Pull request by Zij-IT**
Here is a quick rundown of the changes:

- Additions:
    - `zero_copy::text::whitespace`
    - `zero_copy::text::newline`
    - `zero_copy::text::digits`
    - `zero_copy::text::int`
    - `zero_copy::text::ident`
    - `zero_copy::text::keyword`
    - `zero_copy::Parser::repeated_slice`
    - `zero_copy::Parser::then_slice`
    - `zero_copy::sliced`
    - `zero_copy::sliced::RepeatedSlice`
    - `zero_copy::sliced::ThenSlice`
    - `zero_copy::text::Char::is_valid_ident_start`
    - `zero_copy::text::Char::is_valid_ident_continue`
    - `zero_copy::text::Char::is_digit`
    - `zero_copy::text::Char::is_zero`

Here are some notes that I took while going through this:

- All of the parsers should have the same parsing capability as on the `master` branch, with slightly more return value flexibility.
- I wasn't really sure if the `_slice` methods that I added to `Parser` should be part of another trait or not. 
- The user still has the same amount of flexibility with `digits` and `int`, though they may have to use an extra `map`. They both still work with `Parser::from_str` as they did on `master`!

If there is anything you find odd, or that looks like it could use improvement, please do let me know! Thanks in advance!

### CODE REVIEW

File: src/zero_copy/sliced.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    internal::{Check, Emit, Mode},
+    Located, PResult, Parser,
+};
+
+pub struct RepeatedSlice<A, I: ?Sized, E = (), S = ()> {
+    pub(crate) parser: A,
+    pub(crate) at_least: usize,
+    pub(crate) at_most: Option<usize>,
+    pub(crate) phantom: PhantomData<(E, S, I)>,
+}
```
Comment (zesterer) [inline src/zero_copy/sliced.rs:15]: Out of interest, how does this combinator work? Is it akin to `x.repeated().map_with_slice(|_, s| s)`?

File: src/zero_copy/text.rs [hunk 2]
```diff
@@ -88,3 +130,368 @@ where
...
+    <I as SliceInput>::Slice: 'a,
+    E: Error<I>,
+    S: 'a,
+    C: From<&'a <I as SliceInput>::Slice>,
+{
+    primitive::any()
+        .filter(move |x: &I::Token| x.is_digit(radix))
+        .repeated_slice()
+        .at_least(1)
+        .map(C::from)
```
Comment (zesterer) [inline src/zero_copy/text.rs:205]: I'm not sure I understand what this `C::from` does. How is the conversion occurring, and into what?

**Review by zesterer — COMMENTED** [review]
Thanks for the changes, I really appreciate the time you've put into this, along with the extensive tests :) I just have a few questions about some things.

**Comment by Zij-IT** [comment]
This is just a bump, and something to let you know that I have done some cleanup of the branch since your last comment, and unless I find another bug, or you have any requests regarding this branch, I am done making commits to it :D

File: src/zero_copy/sliced.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    type Output = &'a I::Slice;
+
+    fn go<M: Mode>(&self, inp: &mut InputRef<'a, '_, I, E, S>) -> PResult<M, Self::Output, E> {
+        let before = inp.save();
+        let _a = self.parser_a.go::<M>(inp)?;
+        let _b = self.parser_b.go::<M>(inp)?;
+        let after = inp.save();
+
+        Ok(M::bind(|| inp.slice(before..after)))
+    }
```
Comment (zesterer) [inline src/zero_copy/sliced.rs:138]: This is just equivalent to `a.then(b).map_with_slice(|_, slice| ...)`, no?

**Comment by Zij-IT** [comment]
Hey, I think it's safe to say that this pull request can be closed, as it appears like most (if not all) the important changes have been made to `zero_copy`. If there's something in here that you'd like me to salvage I would be happy to try, otherwise feel free to close the PR :)

**Pull request closed without merging**